### PR TITLE
Ports zig-uri to stdlib.

### PR DESCRIPTION
**Pull request by MasterQ32**
Implements #14176

### CODE REVIEW

File: lib/std/Url.zig [hunk 1]
```diff
@@ -1,98 +1,519 @@
-scheme: []const u8,
-host: []const u8,
-path: []const u8,
+//! Implements URI parsing roughly adhere to https://tools.ietf.org/html/rfc3986
+//! Does not do perfect grammar and character class checking, but should be robust against
+//! "wild" URIs
```
Comment (r00ster91) [inline lib/std/Url.zig:3]:
```suggestion
//! Implements URI parsing roughly adhering to <https://tools.ietf.org/html/rfc3986>.
//! Does not do perfect grammar and character class checking, but should be robust against URIs in the wild.
```

File: lib/std/Url.zig [hunk 2]
```diff
@@ -1,98 +1,519 @@
...
+    for (list) |uri| {
+        _ = try parse(uri);
+    }
+}
+
+test "Special test" {
+    // This is for all of you code readers ♥
+    _ = try parse("https://www.youtube.com/watch?v=dQw4w9WgXcQ&feature=youtu.be&t=0");
+}
+
```
Comment (r00ster91) [inline lib/std/Url.zig:500]:
```suggestion
```

File: lib/std/Url.zig [hunk 3]
```diff
@@ -1,98 +1,519 @@
...
+    };
+    for (list) |uri| {
+        _ = try parse(uri);
+    }
+}
+
+// source:
+// https://tools.ietf.org/html/rfc3986#section-5.4.1
+test "Examples from RFC3986" {
+    // these should all parse
```
Comment (r00ster91) [inline lib/std/Url.zig:466]:
```suggestion
```

File: lib/std/Url.zig [hunk 4]
```diff
@@ -1,98 +1,519 @@
...
+        .path = uri[4..],
+        .query = null,
+        .fragment = null,
+    }, try parse(uri));
+}
+
+// source:
+// https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Examples
+test "Examples from wikipedia" {
+    // these should all parse
```
Comment (r00ster91) [inline lib/std/Url.zig:447]: Seems redundant.
```suggestion
```

File: lib/std/Url.zig [hunk 5]
```diff
@@ -1,98 +1,519 @@
...
+        "a",
+        "a.b",
+        "example.com",
+        "www.example.com",
+        "example.org.",
+        "www.example.org.",
+        "xn--nw2a.xn--j6w193g", // internationalization!
+        "fe80--1ff-fe23-4567-890as3.ipv6-literal.net",
+    });
+    // still allowed…
```
Comment (r00ster91) [inline lib/std/Url.zig:383]:
```suggestion
```

File: lib/std/Url.zig [hunk 6]
```diff
@@ -1,98 +1,519 @@
...
+
+test "authority.dns-names" {
+    try testAuthorityHost(.{
+        "a",
+        "a.b",
+        "example.com",
+        "www.example.com",
+        "example.org.",
+        "www.example.org.",
+        "xn--nw2a.xn--j6w193g", // internationalization!
```
Comment (r00ster91) [inline lib/std/Url.zig:380]:
```suggestion
        "xn--nw2a.xn--j6w193g", // internationalized URI: 見.香港
```

**Review by r00ster91 — CHANGES_REQUESTED** [review]
The TODO said "TODO: redo this implementation according to RFC 1738" and that RFC is the Uniform Resource **Locators** (URL) RFC.
The RFC 3986 implemented here is the Uniform Resource **Identifier** (URI) RFC, so I wonder if this should be renamed to `std.URI`.
There is a difference between a URL and a URI and the library itself seems to use that term too.

A URL is a specific type of URI, so a URI seems more general, so I think it would make sense to rename.

File: lib/std/Url.zig (hunk 2 above)
Comment (MasterQ32) [reply lib/std/Url.zig]: Not sure if i want to remove that. It's a very important URL

File: lib/std/Uri.zig [hunk 7]
```diff
@@ -0,0 +1,515 @@
+//! Implements URI parsing roughly adhering to <https://tools.ietf.org/html/rfc3986>.
+//! Does not do perfect grammar and character class checking, but should be robust against URIs in the wild.
+
+const Url = @This();
```
Comment (r00ster91) [inline lib/std/Uri.zig:4]:
```suggestion
const Uri = @This();
```
After that there should be zero "URL" occurrences.

**Comment by matu3ba** [comment]
Is this post still relevant https://perishablepress.com/stop-using-unsafe-characters-in-urls/#unsafe-characters?
```
A Dangerous Trend

Note: Things have changed a lot with Google and URI specification. So the following information may be outdated, but the example serves to illustrate how developers should be mindful of specification when crafting URIs in applications.
```

Also:
```
Note that RFC 1738 is now obsolete, however the information remains useful in the general sense, and is shared below for educational and reference purposes.
```

File: lib/std/Uri.zig (hunk 7 above)
Comment (MasterQ32) [reply lib/std/Uri.zig:4]: Oh right, good point! :see_no_evil:

**Comment by MasterQ32** [comment]
@matu3ba i never took a look at the older uri specifications, i've made the implementation right after RFC 3986, so it should be up-to-date in those regards. I can double-check later tho

File: lib/std/Uri.zig [hunk 8]
```diff
@@ -0,0 +1,515 @@
+//! Implements URI parsing roughly adhering to <https://tools.ietf.org/html/rfc3986>.
+//! Does not do perfect grammar and character class checking, but should be robust against URIs in the wild.
+
+const Uri = @This();
+const std = @import("std.zig");
+const testing = std.testing;
+
+scheme: ?[]const u8,
```
Comment (truemedian) [inline lib/std/Uri.zig:8]: scheme is not optional, all forms of URI require a scheme (just like they require a path)

**Review by truemedian — CHANGES_REQUESTED** [review]

File: lib/std/Uri.zig (hunk 8 above)
Comment (andrewrk) [reply lib/std/Uri.zig:8]: what about host?

Comment (truemedian) [reply lib/std/Uri.zig:8]: URI encompass URNs, which have no authority. The most simple URI is `scheme:something` where `something` is the "path"

File: lib/std/Url.zig (hunk 2 above)
Comment (AndersonTorres) [reply lib/std/Url.zig]: Rick-rolled by Nixpkgs! OH NO!

**Pull request merged**